### PR TITLE
[SPARK-31730][CORE][TEST] Fix flaky tests in BarrierTaskContextSuite

### PR DESCRIPTION
**Pull request by jiangxb1987**
### What changes were proposed in this pull request?

To wait until all the executors have started before submitting any job. This could avoid the flakiness caused by waiting for executors coming up.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you so much, @jiangxb1987 .

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by Ngone51 — APPROVED** [review]
LGTM

**Comment by jiangxb1987** [comment]
"SPARK-31485: barrier stage should fail if only partial tasks are launched" This test case is flaky, @Ngone51 can you take a look?

**Comment by Ngone51** [comment]
sure, thanks!

**Comment by Ngone51** [comment]
retest this please.

**Comment by WeichenXu123** [comment]
Also fix this test like:
```
  test("share messages with allGather() call") {
    val conf = new SparkConf()
      .setMaster("local-cluster[4, 1, 1024]")
      .setAppName("test-cluster")
    sc = new SparkContext(conf)
    val rdd = sc.makeRDD(1 to 4, 4)
    val rdd2 = rdd.barrier().mapPartitions { it =>
      val context = BarrierTaskContext.get()
      // Sleep for a random time before global sync.
      Thread.sleep(Random.nextInt(1000))
      // Pass partitionId message in
      val message: String = context.partitionId().toString
      val messages: Array[String] = context.allGather(message)
      Iterator.single(messages.toList)
    }
    // Take a sorted list of all the partitionId messages
    val messages_list = rdd2.collect()
    assert (messages_list.length === 4)
    for (messages <- messages_list) {
      assert (messages === (0 until 4).map(_.toString).toList)
    }
  }
```

**Comment by Ngone51** [comment]
Opened a separate PR https://github.com/apache/spark/pull/28596 to fix the test @WeichenXu123

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Since #28596 is merged, shall we restart to proceed this PR? Thanks.

File: core/src/test/scala/org/apache/spark/scheduler/BarrierTaskContextSuite.scala [hunk 1]
```diff
@@ -37,10 +37,10 @@ class BarrierTaskContextSuite extends SparkFunSuite with LocalSparkContext with
       .setAppName("test-cluster")
       .set(TEST_NO_STAGE_RETRY, true)
     sc = new SparkContext(conf)
+    TestUtils.waitUntilExecutorsUp(sc, numWorker, 60000)
```
Comment (seanli-rallyhealth) [inline core/src/test/scala/org/apache/spark/scheduler/BarrierTaskContextSuite.scala:40]: can 60000 be configurable?

**Review by seanli-rallyhealth — APPROVED** [review]

**Comment by Ngone51** [comment]
ping @jiangxb1987 Could you please update this?

**Comment by Ngone51** [comment]
retest this please.

**Comment by Ngone51** [comment]
retest this please.

**Comment by jiangxb1987** [comment]
Retest this please.

**Comment by jiangxb1987** [comment]
Retest this please.

**Comment by jiangxb1987** [comment]
Retest this please.

**Comment by jiangxb1987** [comment]
Let's trigger one more time to ensure the suite is no longer flaky.

**Comment by jiangxb1987** [comment]
Retest this please.

**Comment by jiangxb1987** [comment]
Thanks, merged to master/3.0 !

**Comment by jiangxb1987** [comment]
Opened https://github.com/apache/spark/pull/28658 for branch-3.0

**Pull request closed without merging**